core/src/main/scala/za/co/absa/fadb/DBEngine.scala

```diff
@@ -16,57 +16,69 @@
 
 package za.co.absa.fadb
 
-import scala.concurrent.{ExecutionContext, Future}
+import cats.Monad
+import cats.implicits.toFunctorOps
+import za.co.absa.fadb.exceptions.StatusException
+
 import scala.language.higherKinds
 
 /**
-  * A basis to represent a database executor
-  */
-trait DBEngine {
+ *  `DBEngine` is an abstract class that represents a database engine.
+ *  It provides methods to execute queries and fetch results from a database.
+ *  @tparam F - The type of the context in which the database queries are executed.
+ */
+abstract class DBEngine[F[_]: Monad] {
 
   /**
-    * A type representing the (SQL) query within the engine
-    * @tparam T - the return type of the query
-    */
-  type QueryType[T] <: Query[T]
+   *  A type representing the (SQL) query within the engine
+   *  @tparam R - the return type of the query
+   */
+  type QueryType[R] <: Query[R]
+  type QueryWithStatusType[R] <: QueryWithStatus[_, _, R]
```

Comment:
perhaps these two can even live in `Query.scala`

Reply:
I believe they have to be defined like this in DBEngine, otherwise we wouldn't be able to reference them as dBEngine.QueryType[R] and dBEngine.QueryWithStatusType[R] in DBFunction, respectively DBFunctionWithStatus.

Reply:
`QueryType` is a type member of the `DBEngine` abstract class. It's a higher-kinded type, meaning it's a type that takes another type as a parameter. 

The `<: Query[R]` part is a type bound. It means that whatever type `QueryType[R]` is, it must be a subtype of `Query[R]`. `Query[R]` is presumably another type that represents a database query that returns results of type `R`.

In the context of `DBEngine`, `QueryType` is used to abstract over the specific type of queries that a particular database engine can execute. Different implementations of `DBEngine` can use different types of queries, as long as they are subtypes of `Query[R]`.

For example, one implementation of `DBEngine` might use SQL queries, while another might use NoSQL queries. These different types of queries can have different methods and properties, but as long as they are subtypes of `Query[R]`, they can be used with `DBEngine`.

This is a powerful feature of Scala's type system that allows for a high level of abstraction and flexibility in your code. One of the key features of type members is that they can be overriden in subclasses or subtraits.